rpcs3/Emu/Cell/Modules/sceNp.cpp

```diff
@@ -609,14 +608,68 @@ error_code sceNpDrmExecuteGamePurchase()
 
 error_code sceNpDrmGetTimelimit(vm::cptr<char> path, vm::ptr<u64> time_remain)
 {
-	sceNp.todo("sceNpDrmGetTimelimit(path=%s, time_remain=*0x%x)", path, time_remain);
+	sceNp.warning("sceNpDrmGetTimelimit(path=%s, time_remain=*0x%x)", path, time_remain);
 
 	if (!path || !time_remain)
 	{
 		return SCE_NP_DRM_ERROR_INVALID_PARAM;
 	}
 
-	*time_remain = SCE_NP_DRM_TIME_INFO_ENDLESS;
+	//*time_remain = SCE_NP_DRM_TIME_INFO_ENDLESS;
+
+	vm::var<s64> sec;
+	vm::var<s64> nsec;
+
+	// Get system time (real or fake) to compare to
+	error_code ret = sys_time_get_current_time(sec, nsec);
+	if (ret != CELL_OK)
+		return ret;
+
+	const std::string enc_drm_path(path.get_ptr(), std::find(path.get_ptr(), path.get_ptr() + 0x100, '\0'));
+	const auto [fs_error, ppath, real_path, enc_file, type] = lv2_file::open(enc_drm_path, 0, 0);
+
+	if (fs_error)
+	{
+		return {fs_error, enc_drm_path};
+	}
+
+	u32 magic;
+	NPD_HEADER npd;
+
+	enc_file.read<u32>(magic);
+	enc_file.seek(0);
+
+	// Read expiration time from NPD header which is Unix timestamp in milliseconds
+	if (magic == "SCE\0"_u32)
+	{
+		if (!get_npdrm_self_header(enc_file, npd))
+		{
+			sceNp.error(u8"sceNpDrmGetTimelimit(): Failed to read NPD header from sce file '%s'", enc_drm_path);
+			return {SCE_NP_DRM_ERROR_INVALID_PARAM, enc_drm_path};
+		}
+	}
+	else if (magic == "NPD\0"_u32)
+	{
+		// edata / sdata files
+		EDAT_HEADER edat;
+		read_npd_edat_header(&enc_file, npd, edat);
+	}
+	else
+	{
+		// Unknown file type
+		return {SCE_NP_DRM_ERROR_INVALID_PARAM, enc_drm_path};
+	}
+
+	// Convert milliseconds to seconds
+	s64 expire_time = npd.expire_time / 1000ll;
+
+	// Return the remaining time in ticks
+	if (npd.expire_time == 0)
+		*time_remain = SCE_NP_DRM_TIME_INFO_ENDLESS; // TODO
+	else if (*sec >= expire_time)
+		*time_remain = 0;
```

Comment:
```suggestion
	else if (*sec < activation_time || *sec >= expire_time)
		return SCE_NP_DRM_ERROR_TIME_LIMIT;
```

Reply:
it is possible that these 2 errors are also returned in:
sceNpDrmIsAvailable
sceNpDrmIsAvailable2
sceNpDrmProcessExitSpawn2

But I guess this should not be part of this PR

Reply:
Hmm, are you positive this function checks activation time?

Reply:
Returning SCE_NP_DRM_ERROR_SERVICE_NOT_STARTED error produces "There are 20 days remaining in this demo" message in Sonic Generations Demo while returning SCE_NP_DRM_ERROR_TIME_LIMIT does correctly correctly show "This demo has expired" message.

Reply:
Remember, the game-devs decide what to do with the error.
The game is probably just doing something like this:
```
int time_remaining = sceNpDrmGetTimeLimit();

if (time_remaining == SCE_NP_DRM_ERROR_SERVICE_NOT_STARTED)
{
    const int max_demo_time = 20daysInMicroseconds;
    time_remaining = max_demo_time;
}
else if (time_remaining == SCE_NP_DRM_ERROR_TIME_LIMIT)
{
    show_message("This demo has expired");
    return false;
}
else if (time_remaining == Other error)
{
    // some other error handling
}
std::string msg = string_for_remaining_time(time_remaining);
show_message(msg);
return true;
```

Reply:
Yeah, that's entirely possible, could also be that they didn't even bother to to implement a case SCE_NP_DRM_ERROR_SERVICE_NOT_STARTED error.

Reply:
how would they know that there are 20 days remaining otherwise? They must have checked it.